### PR TITLE
JBPM-6587: Stunner - Add Error End Event

### PR DESCRIPTION
**Pull request by wmedvede**

### CODE REVIEW

**Comment by wmedvede** [comment]
Hello  @romartin @hasys , would you mind review this PR?
Thanks.

**Comment by hasys** [comment]
Jenkins execute full downstream build

**Comment by wmedvede** [comment]
This second build failed due to:

ERROR] Failed to execute goal org.codehaus.mojo:gwt-maven-plugin:2.8.1:compile (gwt-compile) on project kie-drools-wb-webapp: Command failed with status 1 -> [Help 1]
11:17:48 org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.codehaus.mojo:gwt-maven-plugin:2.8.1:compile (gwt-compile) on project kie-drools-wb-webapp: Command failed with status 1
11:17:48     at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:213)
11:17:48     at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:154)
11:17:48     at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:146)
11:17:48     at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
11:17:48     at org.apache.maven.lifecycle.internal.builder.multithreaded.MultiThreadedBuilder$1.call

**Comment by wmedvede** [comment]
Jenkins execute full downstream build

**Comment by wmedvede** [comment]
This PR needs to be rebased with https://github.com/kiegroup/kie-wb-common/pull/1296, so don't worry by now.

File: kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/dataproviders/ProcessErrorRefProvider.java [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+            return obj1.toString().compareTo(obj2.toString());
+        }
+    }
+
+    private Collection<String> findCurrentErrorRefs() {
+        Diagram diagram = sessionManager.getCurrentSession().getCanvasHandler().getDiagram();
+        final Set<String> errorRefs = new HashSet<>();
+        @SuppressWarnings("unchecked")
+        final Iterable<Element> it = diagram.getGraph().nodes();
+        it.forEach(element -> {
```
Comment (romartin) [inline kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/dataproviders/ProcessErrorRefProvider.java]: well I'm a big fan of functional programming where possible, and this code seems a good candidate for, isn't it? In addition the use of streams and their intermediate operations usually results in a better performance... 

So probably here you can use an infinite stream and apply some intermediate `filter` / `map` operations to achieve same results, for example:

```
return StreamSupport.stream(it.spliterator(),
                             false)
                .filter(e -> e.getContent() instanceof View)
                .filter(e -> ((View) e.getContent()).getDefinition() instanceof EndErrorEvent)
                .map(e -> ((View) e.getContent()).getDefinition())
                .map(e -> ((EndErrorEvent)e).getExecutionSet().getErrorRef())
                .filter(errorRef -> errorRef != null && errorRef.getValue() != null && !errorRef.getValue().isEmpty())
                .map(ErrorRef::getValue)
                .collect(Collectors.toSet());
```
Even you can map the result to a `Map<Object, String>` as the caller for this method expects, that's up to you.
WDYT?

Comment (romartin) [reply kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/dataproviders/ProcessErrorRefProvider.java]: probably my example above can be reduced a bit more

**Comment by romartin** [comment]
Hey @wmedvede , ready for rebasing? :) Lemme know if I can help. Thanks!

**Comment by wmedvede** [comment]
@romartin @hasys Rebased!
Thanks.

**Comment by hasys** [comment]
Some problems with parent POM file. Fixe already on the way. 

CC @psiroky, @mbiarnes

**Comment by romartin** [comment]
Jenkins please retest this

**Comment by romartin** [comment]
Jenkins please retest this

**Comment by wmedvede** [comment]
last known error not related with this PR

org.kie.workbench.common.services.backend.validation.asset.ValidatorTest.testValidateWhenTheresNoProject

java.lang.NullPointerException: null
	at org.kie.workbench.common.services.backend.validation.asset.ValidatorTest.tearDown(ValidatorTest.java:70)

**Comment by wmedvede** [comment]
let's give it another chance

Jenkins please retest this

**Comment by hasys** [comment]
Jenkins execute full downstream build

**Comment by wmedvede** [comment]
this time failed due to 

org.kie.wb.test.rest.functional.OrganizationalUnitIntegrationTest.org.kie.wb.test.rest.functional.OrganizationalUnitIntegrationTest

javax.ws.rs.InternalServerErrorException: HTTP 500 Internal Server Error
	at org.jboss.resteasy.client.jaxrs.internal.ClientInvocation.handleErrorStatus(ClientInvocation.java:209)
	at org.jboss.resteasy.client.jaxrs.internal.ClientInvocation.extractResult(ClientInvocation.java:174)

**Comment by wmedvede** [comment]
Jenkins please retest this

**Comment by wmedvede** [comment]
again...
org.kie.wb.test.rest.functional.OrganizationalUnitIntegrationTest.org.kie.wb.test.rest.functional.OrganizationalUnitIntegrationTest
avax.ws.rs.InternalServerErrorException: HTTP 500 Internal Server Error
	at org.jboss.resteasy.client.jaxrs.internal.ClientInvocation.handleErrorStatus(ClientInvocation.java:209)

**Comment by wmedvede** [comment]
same failure again
org.kie.wb.test.rest.functional.OrganizationalUnitIntegrationTest.org.kie.wb.test.rest.functional.OrganizationalUnitIntegrationTest

HTTP 500 Internal Server Error
Stacktrace

javax.ws.rs.InternalServerErrorException: HTTP 500 Internal Server Error

**Comment by wmedvede** [comment]
Jenkins please retest this

**Comment by wmedvede** [comment]
This was the error this time

org.kie.workbench.common.forms.editor.backend.indexing.FormDefinitionIndexerTest.testFormsIndexing
java.lang.AssertionError: expected:<5> but was:<6>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:834)
	at org.junit.Assert.assertEquals(Assert.java:645)
	at org.junit.Assert.assertEquals(Assert.java:631)
	at org.kie.workbench.common.forms.editor.backend.indexing.FormDefinitionIndexerTest.testFormsIndexing(FormDefinitionIndexerTest.java:138)

**Comment by wmedvede** [comment]
Jenkins please retest this

**Comment by wmedvede** [comment]
and again...

org.kie.wb.test.rest.functional.OrganizationalUnitIntegrationTest.org.kie.wb.test.rest.functional.OrganizationalUnitIntegrationTest

HTTP 500 Internal Server Error
Stacktrace

javax.ws.rs.InternalServerErrorException: HTTP 500 Internal Server Error

**Comment by wmedvede** [comment]
Jenkins please retest this

**Comment by romartin** [comment]
Hey @hasys 
We need you approval for merging this, I just realized that even the random tests failures on the downstream build are present, it seems the WARs are being generated as well, so can you please consume it and do your tests? Once you're happy I'll merge.
CC @wmedvede 
Thanks!

**Comment by romartin** [comment]
sorry @wmedvede it seems you'll have to rebase again... :( The PR was not approved by @hasys yet so I merged the @alepintus one, which is causing issues here.

**Comment by wmedvede** [comment]
re-re-re-rebased

**Comment by romartin** [comment]
Jenkins please retest this

**Comment by wmedvede** [comment]
this was the error this time: (random failing indexing)
org.kie.workbench.common.stunner.bpmn.backend.service.indexing.BpmnFileIndexerTest.testBpmnIndexing
java.lang.AssertionError: expected:<7> but was:<8>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:834)

**Comment by wmedvede** [comment]
Jenkins please retest this

**Comment by hasys** [comment]
Jenkins execute full downstream build

**Comment by wmedvede** [comment]
Number of docs fulfilling the given query criteria expected:<2> but was:<3>
Stacktrace

java.lang.AssertionError: Number of docs fulfilling the given query criteria expected:<2> but was:<3>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:834)
	at org.junit.Assert.assertEquals(Assert.java:645)
	at org.kie.workbench.common.services.refactoring.backend.server.IndexingTest.searchFor(IndexingTest.java:240)

**Comment by hasys** [comment]
Hi @wmedvede, Full downstream build is failing due to some issues in drools-wb, I am not able to test binaries locally now.

[ERROR] /home/jenkins/workspace/jbpm-designer-downstream-pullrequests/upstream-repos/drools-wb/drools-wb-webapp/src/main/java/org/drools/workbench/client/home/HomeProducer.java:[29,1] cannot find symbol

Hi @dgutierr is it possible that this issues is related to: https://github.com/kiegroup/drools-wb/pull/722 ?

**Comment by dgutierr** [comment]
@hasys Not sure. The PR you mention has been merged a few minutes ago. It could be a temporal build failure due to multiple PRs getting merged. Let's give it another try...

Jenkins execute full downstream build

**Comment by romartin** [comment]
Jenkins please retest this

**Pull request merged**